### ISSUE TITLE
memlab API findLeaks sometimes freeze on step calculateDominatorNodesFromPostOrder()

### ISSUE THREAD

**Issue by ibrayevabay** [facebook/memlab#48]
Hi, I'm using memlab@1.1.30. 

Facing issue: tests freezes on particular pages in our app (not everytime). I suppose that problem related to some objects in snapshot that couldn't parse by memlab. 

In memlab code (step where freezes): 
[/packages/core/src/paths/TraceFinder.ts#L582](https://github.com/facebookincubator/memlab/blob/df78f1123c971ba46275fac5d0acaf4e58b0513c/packages/core/src/paths/TraceFinder.ts#L582)

Provided in attach zip with my data/cur files for failed test.

Thanks in advance for any help!

**Comment by JacksonGL** [comment]
Thanks for reporting this issue with the heap snapshots. For a temporarily fix to unblock you, please add the following code snippet before [this line](https://github.com/facebook/memlab/blob/df78f1123c971ba46275fac5d0acaf4e58b0513c/packages/core/src/paths/TraceFinder.ts#L231):
```
if (postOrderIndex >= nodeCount) {
  postOrderIndex--;
}
```
After applying the temp fix, it should unfreeze the execution. The memory leaks detected will still be correct, but the retained size may be inaccurate.

I need to spend sometime to investigate the heap snapshot and come up with a proper fix.

**Comment by JacksonGL** [comment]
@ibrayevabay Proper fix [here](https://github.com/facebook/memlab/blob/main/packages/core/src/lib/Utils.ts#L951)

`return !!referrer;` --> `return !referrer;`

Commit on the way, may take a while (internal approving + sync to GitHub).

![](https://pbs.twimg.com/media/FfAaOvsXoAA7Vuy.jpg)

**Comment by ibrayevabay** [comment]
@JacksonGL Hi, thanks for the response and quick resolve this issue, very appreciate it <3 
It's not blocking me right now, so will wait new version with fixes, thanks!

**Comment by ibrayevabay** [comment]
Hi. could u please update me when fix will be released. Issue closed, so I expected fixes to be in 1.1.33 version. But still not fixed there:
 
<img width="1477" alt="Screen Shot 2022-10-26 at 17 27 58" src="https://user-images.githubusercontent.com/19988967/198015437-6203af40-19b1-4f4f-89d9-aa86cce90fb8.png">

**Comment by JacksonGL** [comment]
Hi @ibrayevabay, the fix is already in GitHub and should have been included in memlab@1.1.33:
https://github.com/facebook/memlab/blob/dc2b0f43371fb1ac1c59e377d4f88497f5883085/packages/core/src/lib/Utils.ts#L947

**Comment by ibrayevabay** [comment]
Hmmm, very strange. I reinstalled lib 3 times. Cleaned yarn cache, installed everything from scratch, and still see in my project node_modules old implementation:
```
function hasOnlyWeakReferrers(node) {
    const referrer = node.findAnyReferrer((edge) => edge.type !== 'weak' && edge.type !== 'shortcut');
    return !!referrer;
}
```

Could u please recheck, may be this commit not included in 1.1.33?  Thanks in advance!

**Comment by JacksonGL** [comment]
@ibrayevabay I created a test project, npm installed memlab and the fix was included in the installed package.
Make sure you have deleted `package-lock.json`. Also double check the version of `memlab` (1.1.33) and `memlab@core` (1.1.16) in the `node_modules` directory.
